### PR TITLE
HDRP/Enlighten Realtime GI support

### PR DESCRIPTION
**Pull request by KEngelstoft**
### Purpose of this PR
Reactivate Enlighten Realtime GI support for HDRP. Makes it easier to upgrade from builtin render pipeline or HDRP 7 and older.

### Testing status
Manual testing by the developer using Windows10 and a Threadripper 3990X engineering sample, using the template scene with lights and emissive objects tweaked to generate realtime bounce lighting. SRP batcher has been tested.
rtgi-hdrp-emission

Material inspector has been streamlined to work like the build in one (user can now choose between baked and realtime GI emission).

Tech art validation pass done here https://jira.unity3d.com/browse/GFXGI-400

Automated test coverage has been added in the following scenes:
2101_GI_Metapass
2102_GI_Emission
2106_GI_EmissionSG
Realtime GI has been disabled in all other scenes.

Documentation has been updated.

Screenshots, native PRs and more info here https://docs.google.com/document/d/1K6uE2AtJ2lVW2JbM_xO5dFvbR8o2TdXNARJAlzbzYa8/edit#

### Comments to reviewers
Jira task https://jira.unity3d.com/browse/GFXGI-395

### CODE REVIEW

**Review by iM0ve — CHANGES_REQUESTED** [review]
Review: **In Progress**

**Concerns:**
1) After Realtime GI was enabled and Baked was disabled in the Lighting settings. Existing mixed lights show a warning that "Realtime" mode is being used instead of Mixed. However, GI bounce is not visible until you manually switch the mode to Realtime. Is this an issue? Should we update the behavior or the warning text?

https://user-images.githubusercontent.com/36502659/107234252-2530ea00-6a2c-11eb-88b4-5bd78b2e686d.mp4


2) Is it expected that the property is labeled as Deprecated?

image


3) Adding Area light doesnt work like it used to. First part of the video shows expected behaviour, after adding area light you can instantly adjust the indirect intensity. Pont and Spot lights work as expected however.

https://user-images.githubusercontent.com/36502659/107362611-a5655700-6ae1-11eb-8a14-f0bf14d9362c.mp4



**What's tested:**
* Upgrading an old HDRP(4.1) project with Enligten to see if after an upgrade it's still possible to achieve the same result
* Scene debug modes (Albedo, Emissive, Indirect, Directionality)
* Testing with different light types (Point, Spot, Area)
* Day and Night cycle covered by  David Llewelyn [here](https://jira.unity3d.com/browse/GFXGI-400)

**Upgrade result**
Note it is expected that the image is slightly different because in the original, the old Procedural sky was used + old post processing stack. In the upgraded version physically-based sky was used.

Original {HDRP 4.1}

Original

Upgraded {PR branch}

ReEnlighten

**Comment by KEngelstoft** [comment]
> Review: **In Progress**
> 
> **Concerns:**
> 
> 1. After Realtime GI was enabled and Baked was disabled in the Lighting settings. Existing mixed lights show a warning that "Realtime" mode is being used instead of Mixed. However, GI bounce is not visible until you manually switch the mode to Realtime. Is this an issue? Should we update the behavior or the warning text?
> 
>  Cq8CtX4mSJ.mp4 
> 
> 1. Is it expected that the property is labeled as Deprecated?
> 
> image
> 
> 1. Adding Area light doesn't work like it used to. First part of the video shows expected behavior, after adding area light you can instantly adjust the indirect intensity. Point and Spotlights work as expected however.
> 
>  G4J5ccaTmD.mp4 
> 
> 1. New light types such as Pyramid, Box, etc. Also do not work with Enlighten. Not sure if we should add a warning or address this in any way
> 
>  dxOrsjBoxT.mp4 
> 
> **What's tested:**
> 
> * Upgrading an old HDRP(4.1) project with Enlighten to see if after an upgrade it's still possible to achieve the same result
> * Enlighten scene debug modes (Albedo, Emissive, Indirect, Directionality)
> * Different light types (Point, Spot, Area)
> * Checked that real-time emissives work by changing intensity and observing indirect light changes in the Tunnel scene.
> * Day and Night cycle covered by  David Llewelyn [here](https://jira.unity3d.com/browse/GFXGI-400)
> 
> **Upgrade result**
> Note it is expected that the image is slightly different because in the original, the old Procedural sky was used + old post processing stack. In the upgraded version physically-based sky was used.
> 
> Original {HDRP 4.1}
> 
> Original
> 
> Upgraded {PR branch}
> 
> ReEnlighten

1) This behaviour is also present in the built-in render pipeline. I will consult with @w1shmast3r if this should be reported or by design. In any case, it is unrelated to the changes in this PR and will likely have to be addressed in native code if we decide to fix it.

2) Yes, if you are using an older version of the Editor. The label was removed here https://ono.unity3d.com/unity/unity/pull-request/119636/_/lighting/remove-enlighten-rtgi-deprecation (landed in trunk already).

3) It is hard to tell from the video but is the area light intensity so low that it isn't visible?

4) This is by design. Yes, there should be a warning or they should be greyed out in the inspector. I have added a note here so it is not forgotten https://jira.unity3d.com/browse/GFXGI-395

**Comment by iM0ve** [comment]
Updated my review.
To clarify the 3rd concern, Area light GI bounce was working in 4.1 and no longer works in the PR version. In the video I try to highlight this by adjusting the indirect slider

**Comment by KEngelstoft** [comment]
> Updated my review.
> To clarify the 3rd concern, Area light GI bounce was working in 4.1 and no longer works in the PR version. In the video I try to highlight this by adjusting the indirect slider

In HDRP 4.1 Rectangle lights were represented by point lights, this is why you are seeing bounce from the area light.

com.unity.render-pipelines.high-definition\Editor\Lighting\HDLightEditor.cs line 539:
case LightShape.Rectangle:
                    // TODO: Currently if we use Area type as it is offline light in legacy, the light will not exist at runtime
                    //m_BaseData.type.enumValueIndex = (int)LightType.Rectangle;
                    // In case of change, think to update InitDefaultHDAdditionalLightData()
                    settings.lightType.enumValueIndex = (int)LightType.Point;

File: com.unity.render-pipelines.high-definition/Runtime/Material/BuiltinGIUtilities.hlsl [hunk 1]
```diff
@@ -108,7 +108,7 @@ void EvaluateLightmap(float3 positionRWS, float3 normalWS, float3 backNormalWS,
         TEXTURE2D_ARGS(unity_DynamicDirectionality, samplerunity_DynamicLightmap),
         uvDynamicLightmap, unity_DynamicLightmapST, normalWS, backNormalWS, false, decodeInstructions, bakeDiffuseLighting, backBakeDiffuseLighting);
 #else
-    float3 illuminance += SampleSingleLightmap(TEXTURE2D_ARGS(unity_DynamicLightmap, samplerunity_DynamicLightmap), uvDynamicLightmap, unity_DynamicLightmapST, false, decodeInstructions);
+    float3 illuminance = SampleSingleLightmap(TEXTURE2D_ARGS(unity_DynamicLightmap, samplerunity_DynamicLightmap), uvDynamicLightmap, unity_DynamicLightmapST, false, decodeInstructions);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/BuiltinGIUtilities.hlsl:111]: I will do a separate PR for this fix as it need to be backported, thanks for finding it

**Comment by sebastienlagarde** [comment]
Hey, from the latest comment I don't understand if Area light are working or not, could you confirm?

>Area light GI bounce was working in 4.1
Do you mean HDRP 4.1? Don't think we can do any comparison with that anymore. 

If I get the issue, it is about rectangle light not working correctly with real time enlighten?  And it is because at runtime we approximate with a point light right? . I will expect that it is expected to not works as builtin wasn't supporting rect area light at runtime, so the support have never been added for Real time Enlighten right?

**Comment by sebastienlagarde** [comment]
Otherwise the PR is good for me

**Comment by sebastienlagarde** [comment]
Note: could you merge latest master? the PR is really far behing the mainline. Thanks

**Comment by sebastienlagarde** [comment]
Also not sure if we will conflict with this PR: https://github.com/Unity-Technologies/Graphics/pull/3371/files or if the test of this PR will need to be update to not used realtime GI?

**Comment by KEngelstoft** [comment]
> Hey, from the latest comment I don't understand if Area light are working or not, could you confirm?
> 
> > Area light GI bounce was working in 4.1
> > Do you mean HDRP 4.1? Don't think we can do any comparison with that anymore.
> 
> If I get the issue, it is about rectangle light not working correctly with real time enlighten? And it is because at runtime we approximate with a point light right? . I will expect that it is expected to not works as builtin wasn't supporting rect area light at runtime, so the support have never been added for Real time Enlighten right?

Rectangular area lights are not passed as inputs to Enlighten in the native code at this point, but could probably be added later if desired.

**Comment by KEngelstoft** [comment]
@iM0ve I have added warnings in the inspector for HDRP specific light types Pyramid, Box and Tube because they do not work with Enlighten.
inspector

**Comment by iM0ve** [comment]
Will do one last check after the PR is updated with master. 
Also to clarify, I 4th issue was addressed here and also 3rd?

Regarding 3rd, curious why the if editor check was necessary before and is no longer required. Should I do some additional halo testing for this bit?

image

**Comment by KEngelstoft** [comment]
> Will do one last check after the PR is updated with master.
> Also to clarify, I 4th issue was addressed here and also 3rd?
> 
> Regarding 3rd, curious why the if editor check was necessary before and is no longer required. Should I do some additional halo testing for this bit?
> 
> image

The 3rd issue (rect area light not working after upgrade) will be addressed with native code changes here https://ono.unity3d.com/unity/unity/pull-request/120752/_/lighting/enlighten-rect-light

The 4th issue is addressed by the newly pushed changes (https://github.com/Unity-Technologies/Graphics/pull/3385/commits/3696f0ceba14d71729493929e527eed01c8e8fa2) on this branch

**Comment by KEngelstoft** [comment]
You could do additional checks for baked area lights as it looks like they were never using the value from the width / height in HDAdditionalLightData (they used the value from the builtin light data, so often 0.5x0.5). Here is a comparison of before and after for baked lights. This is also fixing it for mixed and realtime area lights with Enlighten.
after
before

**Review by sebastienlagarde — APPROVED** [review]
Thanks for the update! All good for me

**Review by iM0ve — APPROVED** [review]
I upgraded the review and had a last look. All issues were addressed or fixed. Good to go from my side.

**Review by jessebarker — APPROVED** [review]
Approving the Shader Graph change.

**Pull request merged**